File: tools/canonicalize.py
"""Resolve an ingredient string to its canonical Filipino name.

This is the merge key. Two recipes that say "tamarind" and "sampalok mix" are one
purchase, and the grocery list must collapse them to one line before it adds any
quantities up. Everything else in the app depends on this being right.

Pure and dependency-free: give it a string and the synonym map, get a name back.
"""
import json
import pathlib
import re

_WS = re.compile(r"\s+")
_PUNCT = re.compile(r"[^\w\s]")


def normalize(s):
    """Casefold, strip punctuation and collapse whitespace."""
    return _WS.sub(" ", _PUNCT.sub(" ", (s or "").lower())).strip()


def build_index(synonyms):
    """{canonical: [variants]} -> {normalized phrase: canonical}, longest phrase first.

    Longest-first matters: "tamarind mix" must win over "tamarind" so a line that
    says "tamarind mix" is not resolved by the shorter, less specific entry.
    """
    index = {}
    for canonical, variants in synonyms.items():
        for phrase in [canonical, *variants]:
            n = normalize(phrase)
            if n and (n not in index or len(n) > len(normalize(index[n]))):
                index[n] = canonical
    return dict(sorted(index.items(), key=lambda kv: -len(kv[0])))


def resolve(item, synonyms, index=None):
    """-> canonical Filipino name, or None when nothing in the map matches.

    Exact match wins. Otherwise the longest variant that appears as whole words
    inside the string wins, so "young tamarind" and "pork belly, sliced thin"
    still land on sampalok and liempo.
    """
    n = normalize(item)
    if not n:
        return None
    index = index if index is not None else build_index(synonyms)
    if n in index:
        return index[n]
    for phrase, canonical in index.items():
        if re.search(rf"(?<!\w){re.escape(phrase)}(?!\w)", n):
            return canonical
    return None


def load(path=None):
    """Load the shipped synonym map."""
    path = path or pathlib.Path(__file__).resolve().parent.parent / "data" / "synonyms.json"
    return json.loads(pathlib.Path(path).read_text(encoding="utf-8"))

File: src/canonicalize.ts
// Port of tools/canonicalize.py: resolve an ingredient string to its canonical
// Filipino name. This is the merge key, never display text. It reads the same
// data/synonyms.json the Python does; change the map there, not here.

export type Synonyms = Record<string, string[]>;
export type Index = [phrase: string, canonical: string][];

// Python's \w is Unicode-aware, so keep letters, marks and digits in any script.
const PUNCT = /[^\p{L}\p{M}\p{N}_\s]/gu;

/** Casefold, strip punctuation and collapse whitespace. */
export function normalize(s: string | null | undefined): string {
  return (s ?? '').toLowerCase().replace(PUNCT, ' ').replace(/\s+/g, ' ').trim();
}

/** {canonical: [variants]} -> [normalized phrase, canonical][], longest phrase first. */
export function buildIndex(synonyms: Synonyms): Index {
  const index = new Map<string, string>();
  for (const [canonical, variants] of Object.entries(synonyms)) {
    for (const phrase of [canonical, ...variants]) {
      const n = normalize(phrase);
      const prev = index.get(n);
      if (n && (prev === undefined || n.length > normalize(prev).length)) index.set(n, canonical);
    }
  }
  // Stable sort, like Python's sorted(), so equal-length ties keep insertion order.
  return [...index].sort((a, b) => b[0].length - a[0].length);
}

/**
 * -> canonical Filipino name, or null when nothing in the map matches.
 * Exact match wins; otherwise the longest variant found as whole words wins.
 */
export function resolve(item: string | null | undefined, index: Index): string | null {
  const n = normalize(item);
  if (!n) return null;
  const exact = index.find(([phrase]) => phrase === n);
  if (exact) return exact[1];
  // normalize() leaves single-spaced word runs, so padding with spaces gives the
  // same whole-word match as Python's (?<!\w)phrase(?!\w).
  const padded = ` ${n} `;
  const hit = index.find(([phrase]) => padded.includes(` ${phrase} `));
  return hit ? hit[1] : null;
}
